### ISSUE TITLE
BIDMat: Special Sparse Constructor

### ISSUE THREAD

**Issue by ago109** [BIDData/BIDMach#86]
I have a particularly critical use-case where I would like to build a sparse-matrix (SMat, perhaps GSMat later) from the columns I have accessed from an FMat.  The following vignette illustrates how I've achieved this, however, it is costly:

```
var dense_src : FMat = ...(exists from before)...
var col_indices : IMat = ...(exists from before)...
var target = zeros(src.nrows,src.ncols)
target(?,col_indices) = dense_src(?,col_indices)
var result = sparse(target)
```

The sparse(.) function is expensive, plus we are wasting memory by creating a very large "target" of floating-point zeros when all we want is sparse matrix with the columns of floating-point values taken from "dense_src".  Is there any efficient way in BIDMat parlance I may achieve this goal?

Ideally, I'd want the following (but tried and failed, as BIDMat does not support putting in elements that do not pre-exist in a sparse matrix target):

```
var dense_src : FMat = ...(exists from before)...
var col_indices : IMat = ...(exists from before)...
var target = szeros(src.nrows,src.ncols)
target(?,indices) = dense_src(?,col_indices)
```

where "target" never initialized all the un-used zeros and we never have to go from dense to get to sparse.  Alternatively, is there a special SMat constructor that can build from columns of values and an IMat of column indices but NO row indices (i.e., no IMat of row pointers).

Thank you!

**Comment by jcanny** [comment]
Yes, the sparse constructor normally takes row, col, val indices (like matlab). 
If you pass null as the row indices, you get a packed sparse matrix with no internal row index matrix.

val x=sprand(4,4,0.3);full(x)
...
res3: BIDMat.FMat =
  0.16252        0        0        0
  0.36193        0        0        0
        0  0.86685        0        0
  0.13822        0  0.45693        0

val (i,j,v)=find3(x);val y=sparse(null,j,v,4,4);full(y)
...
res5: BIDMat.FMat =
  0.16252  0.86685  0.45693   0
  0.36193        0        0       0
  0.13822        0        0       0

Is that what you're looking for? Also, you *can* actually insert blocks of rows into an existing sparse matrix using the "colslice" method. Most datasources do this in order to reuse the data input matrices and making caching work.

**Comment by ago109** [comment]
Can you please demonstrate what you mean by using colslice?  The vignette I provided above (the latter of the two), returns a "Not Implemented" error.  Looking at Bidmat document, I noticed you cannot "insert" new columns/rows into a sparse matrix unless you already had something in those exact columns/rows before.  I would like to init an empty sparse matrix (as in "szeros(rows,cols)") and specificy which columns to insert the columns I have from the dense FMat ("dense_src" in the example above/earlier).  How would I use "colslice" (in one of its alternative forms, besides the ?/wildcard variant I have in my example) to achieve this?

Also when I try your "sparse(null,j,v,4,4)" I get a null pointer exception, meaning, you have to have row indices...how did you get your constructor to just pack the matrix correctly?

**Comment by jcanny** [comment]
I ran that example form the current master branch. Version 1.03 is getting pretty old. If you pull the code from github and do 
./getdevlibs.sh
you'll get the latest code. 

And you're right, the "norows" version of teh SMat constructor doesnt work with norows. norows is not a widely-used feature and you may have other problems with it. Also for calculations, most routines will add the row indices anyway since they're needed for common sparse matrix kernels. 

To use colslice you do something like:
val a=SMat(3,9,20,0) // makes a sparse matrix which is 3x9 with storage for 20 non-zeros, but empty on creation (last arg=0)
val b = sprand(3,3,0.3);full(b)
        0  0.80420        0
  0.59721  0.55897        0
        0        0  0.52764
 b.colslice(0,3,a,0);  // take rows 0->3 of b, put in "a" starting at col 0
full(a)
        0  0.80420        0        0        0        0        0        0...
  0.59721  0.55897        0        0        0        0        0        0...
        0        0  0.52764        0        0        0        0        0...
b.colslice(0,3,a,3) // take rows 0->3 of b, put in "a" starting at col 3
full(a)
        0  0.80420        0        0  0.80420        0        0        0...
  0.59721  0.55897        0  0.59721  0.55897        0        0        0...
        0        0  0.52764        0        0  0.52764        0        0...
etc.
This will only work if you're adding cols to the end of the sparse matrix a. And you have to have allocated a with enough storage to hold the final result.

**Comment by ago109** [comment]
Ah, I see, so there is no way (in the newest version/update to BIDMat code) to insert into arbitrary columns of an empty sparse matrix?  Much as in the vignette I am wanting to do....

I just noticed creating a dense FMat of zeros is expensive when you're working on the order to let's say a million columns/rows, then setting the appropriate columns of that FMat to what you desire (i.e., select columns of another source FMat) before calling sparse(FMat) on that matrix to make it a proper sparse matrix.  Mind that I am accessing non-contiguous columns via an IMat of indices I compute in a previous step, so the syntax for "matrix(?,indices) = newMatrix(?,indices)" is important...

**Comment by jcanny** [comment]
Sparse matrices are just a compressed form of a triple of vectors with (row, col, value) contents. They're design for efficient aggregate operations, not as constructors. There's no way to do what you want to do with BIDMach sparse type, which is CSC (compressed sparse column), or even with COO (the coordinate form where the column indices are explicit). 

I'd suggest making an Array X of matrices (probably SMats), indexed by columns. Then you'll need to iterate over the columns and copy one sparse column into X. Then walk through X to get all the row, col, value indices into three arrays. You wont need to sort those since the columns and rows will be ordered, but you will need to compress the column indices (SparseMat,compressInds) and make the row indices one-based (SparseMat.incInds). Then you can call the SMat constructor directly with your rows, compressed cols and values arrays). 

Sorry this is so messy. It would be helpful to have an SMatBuilder class for this sort of thing. We havent hit the demand threshold to write it though.
